dotnet/src/Experimental/Process.Core/Internal/KernelProcessStateMetadataExtension.cs

```diff
@@ -0,0 +1,99 @@
+            {
+                builtSteps.Add(step.BuildStep(stepStateObject));
+                return;
+            }
+
+            builtSteps.Add(step.BuildStep());
+        });
+
+        return builtSteps;
+    }
+
+    private static KernelProcessStateMetadata SanitizeProcessStateMetadata(KernelProcessStateMetadata stateMetadata, List<ProcessStepBuilder> stepBuilders)
+    {
+        KernelProcessStateMetadata sanitizedStateMetadata = stateMetadata;
+        stepBuilders.ForEach(step =>
+        {
+            // 1- find matching key name with exact match or by alias match
+            string? stepKey = null;
+
+            if (sanitizedStateMetadata.StepsState != null && sanitizedStateMetadata.StepsState.ContainsKey(step.Name))
+            {
+                stepKey = step.Name;
+            }
+            else
+            {
+                stepKey = step.Aliases
+                    .Where(alias => sanitizedStateMetadata.StepsState != null && sanitizedStateMetadata.StepsState.ContainsKey(alias))
+                    .FirstOrDefault();
+            }
+
+            // 2- stepKey match found
+            if (stepKey != null)
+            {
+                var currentVersionStateMetadata = step.BuildStep().ToProcessStateMetadata();
+                if (sanitizedStateMetadata.StepsState!.TryGetValue(stepKey, out var savedStateMetadata))
+                {
+                    if (stepKey != step.Name)
+                    {
+                        if (savedStateMetadata.VersionInfo == currentVersionStateMetadata.VersionInfo)
+                        {
+                            // key mismatch only, but same version
+                            sanitizedStateMetadata.StepsState[step.Name] = savedStateMetadata;
+                            // TODO: Should there be state formatting check too?
+                        }
+                        else
+                        {
+                            // version mismatch - check if migration logic in place
+                            if (step is ProcessBuilder subprocessBuilder)
+                            {
+                                var sanitizedStepState = SanitizeProcessStateMetadata(savedStateMetadata, subprocessBuilder.Steps.ToList());
+                                sanitizedStateMetadata.StepsState[step.Name] = sanitizedStepState;
+                            }
+                            else if (false)
+                            {
+                                // TODO: Hook up logic for custom step override
+                            }
+                            else
+                            {
+                                // no compatible state found, migrating id only
+                                sanitizedStateMetadata.StepsState[step.Name] = new KernelProcessStateMetadata()
```

Comment:
In this case are we loosing state data? If so, should we through an exception?

Reply:
so the part 3 of these changes is to support custom override/migration so if there was one set that would take over, else only the matching name and id would migrate, and a default step state would be set